pkg/system/phase2_creating.go

```diff
@@ -1482,6 +1493,187 @@ func (r *Reconciler) UpgradeMigrateDB() error {
+	if err := r.ReconcileObject(r.NooBaaPostgresDB, func() error {
+		podSpec := &r.NooBaaPostgresDB.Spec.Template.Spec
+		podSpec.ServiceAccountName = "noobaa"
+		for i := range podSpec.InitContainers {
+			c := &podSpec.InitContainers[i]
+			if c.Name == "upgrade-db" {
+				c.Command = []string{"sh", "-x", initScript}
+				c.Image = targetImage
+			}
+		}
+		for i := range podSpec.Containers {
+			c := &podSpec.Containers[i]
+			if c.Name == "db" {
+				c.Image = targetImage
+			}
+		}
+		return nil
+	}); err != nil {
+		r.Logger.Errorf("got error on postgres STS reconcile %v", err)
+		return err
+	}
+	return nil
+}
+
+// HasUpgradeDbContainerFailed checks if postgres upgrade container failed
+func (r *Reconciler) HasUpgradeDbContainerFailed(dbPod *corev1.Pod) bool {
+	//when init container restart policy is Always, pod state will not be set to failed on container error.
+	//check for persistent error in container
+	for i := range dbPod.Status.InitContainerStatuses {
+		if dbPod.Status.InitContainerStatuses[i].Name == "upgrade-db" {
+			if dbPod.Status.InitContainerStatuses[i].State.Waiting != nil && dbPod.Status.InitContainerStatuses[i].State.Waiting.Reason == "CrashLoopBackOff" {
+				return true
+			}
+			break
+		}
+	}
+	return false
+}
+
+// RevertPostgresUpgrade reverts the postgres version to original version after failed upgrade
+func (r *Reconciler) RevertPostgresUpgrade() error {
+	r.Logger.Infof("RevertPostgresUpgrade reverting postgres upgrade to original database image")
+	if err := r.ReconcileSetDbImageAndInitCode(*r.NooBaa.Status.BeforeUpgradeDbImage, "/init/revertdb.sh"); err != nil {
+		r.Logger.Errorf("got error on postgres STS reconcile %v", err)
+		return err
+	}
+	//override DB image with older image
+	if err := r.ReconcileObject(r.NooBaa, func() error {
+		r.NooBaa.Spec.DBImage = r.NooBaa.Status.BeforeUpgradeDbImage
+		return nil
+	}); err != nil {
+		return err
+	}
+	//since DB pod is in failed state need to restart it
+	restartError := r.RestartDbPods()
+	if restartError != nil {
+		r.Logger.Warn("Unable to restart db pods")
+	}
+	return nil
+}
+
+// UpgradePostgresDB upgrade postgres image to newer verision
+func (r *Reconciler) UpgradePostgresDB() error {
+	phase := r.NooBaa.Status.PostgresUpdatePhase
+	if phase == nbv1.UpgradePhaseFinished || phase == nbv1.UpgradePhaseNone {
+		return nil
+	}
+
+	var err error = nil
+	hasInitContainerFailed := false
+	switch phase {
+
+	case "":
+		sts := util.KubeObject(bundle.File_deploy_internal_statefulset_db_yaml).(*appsv1.StatefulSet)
+		if r.NooBaa.Spec.DBType == "postgres" {
+			sts.Name = "noobaa-db-pg"
+		}
+		sts.Namespace = options.Namespace
+		//postgres database doesn't exists, no need to upgrade.
+		if !util.KubeCheckQuiet(sts) {
+			phase = nbv1.UpgradePhaseNone
+			break
+		}
+
+		oldImage := sts.Spec.Template.Spec.Containers[0].Image
+		if oldImage == GetDesiredDBImage(r.NooBaa) {
```

Comment:
If the upgrade will break after phase `UpgradePhasePrepare`, we will not continue the upgrade in the next `Reconcile`, since the STS image will be the same as desired

Reply:
@dannyzaken, maybe I missed something, but as we discussed and if there is no failthorough, we will first set the phase to prepare before we fail on updating the DB STS, no?